apps/dog-food/ChatTerminal.hs

```diff
@@ -0,0 +1,329 @@
+      iStart = th - ih
+      prompt = inputPrompt ts
+      (cRow, cCol) = relativeCursorPosition tw $ length prompt + inputPosition ts
+  if nmr >= iStart
+    then atomically $ writeTVar nextMessageRow iStart
+    else clearLines nmr iStart
+  C.setCursorPosition (max nmr iStart) 0
+  putStr $ prompt <> inputString ts <> " "
+  C.clearFromCursorToLineEnd
+  C.setCursorPosition (iStart + cRow) cCol
+  C.showCursor
+  where
+    clearLines :: Int -> Int -> IO ()
+    clearLines from till
+      | from >= till = return ()
+      | otherwise = do
+        C.setCursorPosition from 0
+        C.clearFromCursorToLineEnd
+        clearLines (from + 1) till
+
+    relativeCursorPosition :: Int -> Int -> (Int, Int)
+    relativeCursorPosition width pos =
+      let row = pos `div` width
+          col = pos - row * width
+       in (row, col)
+
+updateUsername :: ChatTerminal -> Maybe Contact -> STM ()
+updateUsername ct a = do
+  writeTVar (username ct) a
+  modifyTVar (termState ct) $ \ts -> ts {inputPrompt = promptString a}
+
+promptString :: Maybe Contact -> String
+promptString a = maybe "" (B.unpack . toBs) a <> "> "
+
+sendToTTY :: ChatTerminal -> IO ()
+sendToTTY ChatTerminal {outputQ} =
+  forever $ atomically (readTBQueue outputQ) >>= putLn stdout
+
+sendToTTY' :: ChatTerminal -> IO ()
+sendToTTY' ct@ChatTerminal {outputQ} =
+  forever $ atomically (readTBQueue outputQ) >>= printMessage ct >> updateInput ct
+
+printMessage :: ChatTerminal -> ByteString -> IO ()
+printMessage ChatTerminal {termSize, nextMessageRow} msg = do
+  nmr <- readTVarIO nextMessageRow
+  C.setCursorPosition nmr 0
+  let (th, tw) = termSize
+  lc <- printLines tw msg
+  atomically . writeTVar nextMessageRow $ min (th - 1) (nmr + lc)
+  where
+    printLines :: Int -> ByteString -> IO Int
+    printLines tw s = do
+      let ls
+            | B.null s = [""]
+            | otherwise = B.lines s <> ["" | B.last s == '\n']
+      print_ ls
+      return $ foldl (\lc l -> lc + (B.length l `div` tw) + 1) 0 ls
+
+    print_ :: [ByteString] -> IO ()
+    print_ [] = return ()
+    print_ (l : ls) = do
+      B.hPut stdout l
+      C.clearFromCursorToLineEnd
+      B.hPut stdout "\n"
+      print_ ls
+
+getKey :: IO Key
+getKey = charsToKey . reverse <$> keyChars ""
+  where
+    charsToKey = \case
+      "\ESC" -> KeyEsc
+      "\ESC[A" -> KeyUp
+      "\ESC[B" -> KeyDown
+      "\ESC[D" -> KeyLeft
+      "\ESC[C" -> KeyRight
+      "\ESCb" -> KeyAltLeft
+      "\ESCf" -> KeyAltRight
+      "\ESC[1;5D" -> KeyCtrlLeft
+      "\ESC[1;5C" -> KeyCtrlRight
+      "\ESC[1;2D" -> KeyShiftLeft
+      "\ESC[1;2C" -> KeyShiftRight
+      "\n" -> KeyEnter
+      "\DEL" -> KeyBack
+      "\t" -> KeyTab
+      '\ESC' : _ -> KeyUnsupported
+      cs -> KeyChars cs
+
+    keyChars cs = do
+      c <- getChar
+      more <- hReady stdin
+      -- for debugging - uncomment this, comment line after: 
+      -- (if more then keyChars else \c' -> print (reverse c') >> return c') (c : cs)
```

Comment:
if you don't mind I'll keep this as a comment - I myself criticize commented code, but this will just get squashed and we'll have to recollect it, and it's just very convenient for testing. I think it's okay to keep this one line commented together with the comment explaining it's for debugging